### PR TITLE
Upgrade, expand unit tests

### PR DESCRIPTION
**Pull request by ghostwords**
As followup to #1182.

Progress:

- [x] Upgrade QUnit
- [x] Start faking XMLHttpRequests in unit tests
- [x] Add more DNT tests
- [x] Add some cookie tests (following up on #1259)

### CODE REVIEW

**Comment by cowlicks** [comment]
This is awesome!

Can I suggest that instead of adding Qunit and sinon in `lib/vend` we just list them as dependencies in our `package.json`?

I think it would be good because:
* keeps tons of lines out of version control to keep `git clone` small.
* We don't have to worry about excluding the files in our build process since they'll be in `node_modules` or wherever
* upgrading is easier

File: src/utils.js [hunk 1]
```diff
@@ -96,6 +96,10 @@ function xhrRequest(url, callback, method){
       callback(error, error.message);
     }
   };
+  // triggered by network problems
+  xhr.onerror = function () {
+    callback({ status: 0, message: "", object: xhr }, "");
+  };
```
Comment (cowlicks) [inline src/utils.js]: Are you trying to silence the 404 etc. error messages from checking for DNT policies in the console? If so, bad news, this is deliberately impossible by design (at least in chrome).

Comment (ghostwords) [inline src/utils.js]: No, this is to restore previous `onreadystatechange`-powered behavior; [`onload` doesn't get triggered by network failures, but `onreadystatechange` does](http://stackoverflow.com/a/19247992). You can trigger this scenario by calling `utils.xhrRequest` while offline.

**Comment by cooperq** [comment]
lgtm

**Comment by ghostwords** [comment]
Responding to https://github.com/EFForg/privacybadger/pull/1250#issuecomment-286632862: My concern with moving libraries to package.json is it will add a build step for relatively small gains. To run unit tests, first install `node`, find the extension folder, install dependencies ... all to save a few KB? We may very well do it in the (React-powered) future! Just not feeling it right now.

**Pull request merged**